pyvo/dal/adhoc.py

```diff
@@ -167,48 +170,127 @@ def get_adhocservice_by_id(self, id_):
-        """
-        remaining_ids = []  # remaining IDs to processed
-        current_batch = None  # retrieved but not returned yet
-        current_ids = []  # retrieved but not returned
-        processed_ids = []  # retrived and returned IDs
-        batch_size = None  # size of the batch
 
-        for row in self:
-            if not current_ids:
-                if batch_size is None:
-                    # first call.
-                    self.query = DatalinkQuery.from_resource(
-                        [_ for _ in self],
-                        self._datalink,
-                        session=self._session,
-                        original_row=row)
-                    remaining_ids = self.query['ID']
-                if not remaining_ids:
-                    # we are done
-                    return
-                if batch_size:
-                    # subsequent calls are limitted to batch size
-                    self.query['ID'] = remaining_ids[:batch_size]
-                current_batch = self.query.execute(post=True)
-                current_ids = list(OrderedDict.fromkeys(
-                    [_ for _ in current_batch.to_table()['ID']]))
-                if not current_ids:
-                    raise DALServiceError(
-                        'Could not retrieve datalinks for: {}'.format(
-                            ', '.join([_ for _ in remaining_ids])))
-                batch_size = len(current_ids)
-            id1 = current_ids.pop(0)
-            processed_ids.append(id1)
-            remaining_ids.remove(id1)
-            yield current_batch.clone_byid(
-                id1,
-                original_row=row)
+        Parameters
+        ----------
+
+        preserve_order : bool
+            True to return the datalinks keeping the order of the current rows.
+            NOTE: There might be a performance penalty for keeping the order as
+            one request per row is sent to the service. When the order of the
+            datalinks is not important, the execution is optimized to query the
+            service in batches.
+
+        """
+        def _get_results_tb(rows, dl_batch_tb):
+            # Creates a new DL result table with the given rows as the results data
+            # and the dl_batch_tb as the template for both table fields and other resources
+            tb = VOTableFile()
+            new_table = TableElement(tb)
+            new_table.fields.extend(dl_batch_tb.get_first_table().fields)
+            new_table.create_arrays(len(rows))
+            for index, row in enumerate(rows):
+                new_table.array[index] = row
+            results_resource = Resource()
+            results_resource.type = "results"
+            results_resource.tables.append(new_table)
+            tb.resources.append(results_resource)
+            # now add all the other resources from the dl_batch_tb
+            for resource in dl_batch_tb.resources:
+                if resource.type != "results":
```

Comment:
I think I'd prefer a comparison resource.type=="meta"; I don't think we'll ever have resource types blowing this up in any way, but then comparing for equality we roughly know what we're doing, which we don't when we handle some unknown future RESOURCE type.